packages/devtools_shared/test/server/general_api_test.dart

```diff
@@ -0,0 +1,274 @@
+        () async {
+          final request = Request(
+            'get',
+            Uri(
+              scheme: 'https',
+              host: 'localhost',
+              path: apiNotifyForVmServiceConnection,
+              queryParameters: {
+                apiParameterValueKey: 'fake_uri',
+                apiParameterVmServiceConnected: 'true',
+              },
+            ),
+          );
+          final response = await server.ServerApi.handle(
+            request,
+            extensionsManager: ExtensionsManager(buildDir: '/'),
+            deeplinkManager: FakeDeeplinkManager(),
+            dtd: (uri: 'ws://dtd:uri', secret: 'fake_secret'),
+            analytics: const NoOpAnalytics(),
+          );
+          expect(response.statusCode, HttpStatus.badRequest);
+          final respasString = await response.readAsString();
+          expect(
+            respasString,
+            contains('Cannot normalize VM service URI'),
+          );
+        },
+      );
+      test(
+        'returns badRequest for invalid $apiParameterVmServiceConnected argument',
+        () async {
+          final request = Request(
+            'get',
+            Uri(
+              scheme: 'https',
+              host: 'localhost',
+              path: apiNotifyForVmServiceConnection,
+              queryParameters: {
+                apiParameterValueKey: 'ws://127.0.0.1:8181/LEpVqqD7E_Y=/ws',
+                apiParameterVmServiceConnected: 'bad_arg',
+              },
+            ),
+          );
+          final response = await server.ServerApi.handle(
+            request,
+            extensionsManager: ExtensionsManager(buildDir: '/'),
+            deeplinkManager: FakeDeeplinkManager(),
+            dtd: (uri: 'ws://dtd:uri', secret: 'fake_secret'),
+            analytics: const NoOpAnalytics(),
+          );
+          expect(response.statusCode, HttpStatus.badRequest);
+          expect(
+            await response.readAsString(),
+            contains('Cannot parse $apiParameterVmServiceConnected parameter'),
+          );
+        },
+      );
+    });
+
+    group('updateDtdWorkspaceRoots', () {
+      TestDtdConnectionInfo? dtd;
+      DTDConnection? testDtdConnection;
+
+      setUp(() async {
+        dtd = await startDtd();
+        testDtdConnection =
+            await DartToolingDaemon.connect(Uri.parse(dtd!.uri!));
+      });
+
+      tearDown(() async {
+        await testDtdConnection?.close();
+        dtd?.dtdProcess?.kill();
+        await dtd?.dtdProcess?.exitCode;
+        dtd = null;
+      });
+
+      test('adds and removes workspace roots', () async {
+        var currentRoots =
+            (await testDtdConnection!.getIDEWorkspaceRoots()).ideWorkspaceRoots;
+        expect(currentRoots, isEmpty);
+
+        final rootUri1 = Uri.parse('file:///Users/me/package_root_1');
+        final rootUri2 = Uri.parse('file:///Users/me/package_root_2');
+
+        await server.Handler.updateDtdWorkspaceRoots(
```

Comment:
Nit: consider creating an inline function to wrap this since you're only changing `rootFromVmService` and `connected` for each test case.

```dart
Future<void> updateDtdWorkspaceRoots({required Uri root, required bool connected}) async {
  await server.Handler.updateDtdWorkspaceRoots(
    (uri: dtd!.uri, secret: dtd!.secret),
    rootFromVmService: root,
    connected: connected,
    api: ServerApi(),
  );
}
```

Reply:
done